core/src/main/scala/chisel3/experimental/OpaqueType.scala

```diff
@@ -16,7 +18,28 @@ import chisel3._
   *
   * @see RecordSpec in Chisel's tests for example usage and expected output
   */
-trait OpaqueType { self: Record =>
+trait OpaqueType extends Record {
+
+  override def do_asUInt(implicit sourceInfo: SourceInfo): UInt = {
+    if (errorOnAsUInt) {
+      val coordinates =
+        if (this.binding.exists(_.isInstanceOf[ChildBinding])) {
+          val n = Arg.earlyLocalName(this, includeRoot = false)
+          s"Field '$n' of type "
+        } else {
+          ""
+        }
+
+      Builder.error(s"${coordinates}${this.typeName} does not support .asUInt.")
+    }
+    super.do_asUInt
+  }
+
+  /** If set to true, calling .asUInt on instances of this type will throw an Exception
+    *
+    * Users can override this to increase the "opacity" of their type.
+    */
+  protected def errorOnAsUInt: Boolean = false
```

Comment:
perhaps they could override it with a string message...? so you could say something like

Reply:
I think you lost the second half of that sentence.

We could make the String message customizable, I'm curious what API you have in mind. I think the current error message is pretty good, eg. from the tests (admittedly hard to see since I also moved a bunch, but see the end of `OpaqueTypeSpec.scala`): `"Field '_.bits' of type MaybeNoAsUInt does not support .asUInt."`. That `Arg.earlyLocalName` trick is very much a private API and IMO you wouldn't want to lose that "coordinate" information when overriding the error message.